features/202-session-management/design.md

```diff
@@ -0,0 +1,163 @@
+# Background
+
+Sessions of auth gear should be configurable and manageable.
+
+
+# Objectives
+
+- App Developer/User should be able to manage sessions.
+- App Developer should be able to configure session lifetime.
+- App Developer should be able to configure session token transport.
+
+
+# Proposed Design
+
+## Stateful Session
+
+Sessions would be stateful (i.e. persisted into database), instead of being
+stateless.
+
+Each session would be associated with attributes:
```

Comment:
I agree with these session attributes, good enough for 90% of usage. 

However I think we need a database structure that use something like an append-only log, to log down all session creation, deletion and access attempts.

i.e. So if developer want to, they can provide a feature to show not just the creation IP and last IP, but also all other IP address used (or agent)

We do not need to have the APIs for the logs access yet in version 1 or this spec, but I think we want the DB structure here too. Use append-only log/DB coz scalability is an issue. This feature could be "fast insert and slow read/query" because session access is extremely common and need to be scalable while accessing session historical data is rare in normal usage.

Reply:
I think the feature you want is access logs, and it is out of scope for this feature.

Reply:
Yes and access log was considered the feature in session management, did you look at the wireframe of the ultimate portal we want to build?

Reply:
I didn't look very clearly last time you showed it, do you have a link for the portal design? In my opinion, access log would be implemented as part of audit log feature.

Requiring DB update every request is not that bad, depending on what DB is used. Also, every request need to hit the DB anyway. This would be an implementation detail, but let's say Redis is used to store sessions.

Reply:
It is in basecamp https://3.basecamp.com/3096882/buckets/11477446/uploads/1962515766

We are using postgresql now I think? Yea sorry to go into a bit implementation here but I just wish we don't duplicate effort in architecting and implementation things, so I thought we can either do it now or leave those "last_xxx" to implement with log altogether

Reply:
Yes, but it is intended to switch to use Redis for sessions.

Reply:
Implementing access log later on should not be too difficult/requiring a rewrite, considering Redis has stream type now.

Reply:
Hmmm, okay, but let's try to avoid touching PostgreSQL DB for each request? I know currently it is but it was accepted as an "MVP implementation trade-off" and will need to be fixed some date -- let's not purposefully add another need to be fixed things in future.

Reply:
@kiootic to resolve this comment and make this design more complete, can you add the note here about:

1. The session/device management features won't be completed without an audit log of session data, so the session attributes will be stored in the log (add the feature issue number) in future and consolidated into this API. The current version of API is considered an MVP versions to only allow access to "latest session information" of each user.
1. Mention the session would be persisted in a low-latency horizontally scalable store for scalability reasons (can mentioned redis if it is decided)
1. Mention country is missing which is too important to ignore.